nntrainer/include/network_graph.h

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file    network_graph.h
+ * @date    19 Oct 2020
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   This is Network Graph Class for Neural Network
+ *
+ */
+
+#ifndef __NETWORK_GRAPH_H_
+#define __NETWORK_GRAPH_H_
+#ifdef __cplusplus
+
+#include <iostream>
+#include <layer_internal.h>
+#include <list>
+#include <loss_layer.h>
+#include <memory>
+#include <stack>
+#include <vector>
+
+namespace nntrainer {
+
+struct LayerNode {
+  std::shared_ptr<Layer> layer;
+  unsigned int index;
+};
+
+class NetworkGraph {
+public:
+  NetworkGraph() : num_node(0), def_name_count(0){};
+
+  void addEdge(unsigned int ith, LayerNode node);
+
+  void topologicalSort();
+
+  void addLayerNode(std::shared_ptr<Layer>);
+
+  unsigned int getNumNode() { return num_node; }
+
+  LayerNode &getLayerNode(unsigned int ith);
+
+  LayerNode &getLayerNode(const std::string layer_name);
```

Comment:
Maybe reference operator?
```suggestion
  LayerNode &getLayerNode(const std::string &layer_name);
```